File: python/src/statemachined/device/device_line_monitor.py
# SPDX-License-Identifier: LGPL-3.0-or-later
"""Every line in and out of the port, as it went, kept for a while.

The thing a person reaches for when the layers above have stopped agreeing:
`/api/device/lines` says the valve is line 3 and the valve is not opening, so
what actually went down the wire? This answers that, in the protocol's own
words, with nothing interpreting them.

**Not the trace.** The trace (`state_visit_trace.py`) is the *record*: one entry
per state visit, timestamped in two clocks, written to disk, and joined to
triald's `.tdr` afterwards. This is a *log*: bytes, in order, thrown away when
the ring wraps, and interesting for about as long as somebody is watching it.
Keeping them apart is what stops the record filling with framing noise nobody
will read in six months.

**It is always on**, because the alternative is not. A link fault that happens
once an hour is not reproducible on demand, and a monitor somebody has to enable
first is a monitor that is off when the interesting thing happens. The cost is
one string appended per line -- a few hundred a second at worst, against a
bounded deque -- which is nothing next to the serial port it is describing.

**And it is only a tail.** `MONITORED_LINE_COUNT` lines, then the oldest go. A
consumer that falls behind is told what it missed, exactly as the trace's stream
does, rather than handed a shorter answer that looks complete.
"""

from __future__ import annotations

import threading
import time
from collections import deque
from typing import Any

#: How many lines the ring holds. A graph set upload is a few hundred lines and
#: a heartbeat is two a second, so this is minutes of an idle link or one whole
#: upload with room around it -- which are the two things anybody scrolls back
#: to look at.
MONITORED_LINE_COUNT = 4000

#: Which way a line went. Spelled out rather than "tx"/"rx", because those are
#: only unambiguous once you have decided whose transmit it is.
TO_DEVICE = "to_device"
FROM_DEVICE = "from_device"


class DeviceLineMonitor:
    """The last few thousand lines, both directions, with their arrival times.

    Thread-safe: it is written from whichever thread holds the link -- the
    background reader, and any request thread sending a command -- and read from
    the API's threads. One lock, held across an append to a deque.
    """

    def __init__(self, ring_lines: int = MONITORED_LINE_COUNT):
        self._lines: deque[dict[str, Any]] = deque(maxlen=ring_lines)
        self._lock = threading.Lock()
        #: Monotonic for the life of the daemon, so a client can say where it
        #: got to. The same cursor the trace uses, for the same reason: a
        #: position in a ring is not a position in a log.
        self._next_entry_number = 0

    @property
    def ring_capacity(self) -> int:
        return self._lines.maxlen or 0

    def record(self, direction: str, line: str) -> dict[str, Any]:
        """One line, as it went. Never raises.

        Called from inside `SerialLink`, which is in the path of every command
        and of the scan-driven `visit` stream, so nothing here may fail: a
        monitor that could break the link would be worse than no monitor.
        """
        with self._lock:
            entry = {
                "entry_number": self._next_entry_number,
                "direction": direction,
                "line": line,
                "recorded_host_time": _iso8601_utc(time.time()),
            }
            self._next_entry_number += 1
            self._lines.append(entry)
            return entry

    # ----------------------------------------------------------- queries ---

    def lines_since(self, entry_number: int = 0, limit: int = 500) -> list[dict[str, Any]]:
        with self._lock:
            return [line for line in self._lines if line["entry_number"] >= entry_number][:limit]

    def newest_entry_number(self) -> int:
        with self._lock:
            return self._next_entry_number - 1

    def oldest_entry_number_still_held(self) -> int:
        with self._lock:
            return self._lines[0]["entry_number"] if self._lines else self._next_entry_number

    def has_fallen_out_of_the_ring(self, entry_number: int) -> bool:
        with self._lock:
            if not self._lines:
                return False
            return entry_number < self._lines[0]["entry_number"]

    def __len__(self) -> int:
        with self._lock:
            return len(self._lines)


def _iso8601_utc(unix_seconds: float) -> str:
    seconds = int(unix_seconds)
    microseconds = round((unix_seconds - seconds) * 1_000_000)
    return time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(seconds)) + f".{microseconds:06d}Z"

File: daemon-rs/src/device/device_line_monitor.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Every line in and out of the port, as it went, kept for a while.
//!
//! The thing a person reaches for when the layers above have stopped agreeing:
//! `ReadLines` says the valve is line 3 and the valve is not opening, so what
//! actually went down the wire? This answers that, in the protocol's own words,
//! with nothing interpreting them.
//!
//! **Not the trace.** The trace is the *record*, written to disk and joined to
//! triald's `.tdr`. This is a *log*: lines, in order, thrown away when the ring
//! wraps. **It is always on**, because a link fault that happens once an hour
//! is not reproducible on demand, and a monitor somebody has to enable first is
//! off when the interesting thing happens.

use std::collections::VecDeque;
use std::sync::Mutex;

use serde_json::{json, Value};

use super::device_clock_correlation::iso8601_utc;

/// How many lines the ring holds: minutes of an idle link, or one whole graph
/// set upload with room around it — the two things anybody scrolls back to.
pub const MONITORED_LINE_COUNT: usize = 4000;

struct Held {
    lines: VecDeque<Value>,
    /// Monotonic for the life of the daemon: a position in a ring is not a
    /// position in a log.
    next_entry_number: i64,
}

/// The last few thousand lines, both directions, with their arrival times.
pub struct DeviceLineMonitor {
    held: Mutex<Held>,
    ring_capacity: usize,
}

impl Default for DeviceLineMonitor {
    fn default() -> Self {
        Self::new(MONITORED_LINE_COUNT)
    }
}

impl DeviceLineMonitor {
    pub fn new(ring_lines: usize) -> Self {
        Self {
            held: Mutex::new(Held {
                lines: VecDeque::new(),
                next_entry_number: 0,
            }),
            ring_capacity: ring_lines,
        }
    }

    fn held(&self) -> std::sync::MutexGuard<'_, Held> {
        self.held
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
    }

    pub fn ring_capacity(&self) -> usize {
        self.ring_capacity
    }

    /// One line, as it went. **Never fails**: this is in the path of every
    /// command and of the visit stream, and a monitor that could break the
    /// link would be worse than none.
    pub fn record(&self, direction: &str, line: &str) {
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|since| since.as_secs_f64())
            .unwrap_or(0.0);
        let mut held = self.held();
        let entry = json!({
            "entry_number": held.next_entry_number,
            "direction": direction,
            "line": line,
            "recorded_host_time": iso8601_utc(now),
        });
        held.next_entry_number += 1;
        if self.ring_capacity > 0 {
            if held.lines.len() == self.ring_capacity {
                held.lines.pop_front();
            }
            held.lines.push_back(entry);
        }
    }

    pub fn lines_since(&self, entry_number: i64, limit: usize) -> Vec<Value> {
        self.held()
            .lines
            .iter()
            .filter(|line| number_of(line) >= entry_number)
            .take(limit)
            .cloned()
            .collect()
    }

    pub fn newest_entry_number(&self) -> i64 {
        self.held().next_entry_number - 1
    }

    pub fn oldest_entry_number_still_held(&self) -> i64 {
        let held = self.held();
        held.lines
            .front()
            .map(number_of)
            .unwrap_or(held.next_entry_number)
    }

    /// Whether a cursor is older than anything still held: a consumer that
    /// fell behind is told what it missed, not handed a shorter answer.
    pub fn has_fallen_out_of_the_ring(&self, entry_number: i64) -> bool {
        self.held()
            .lines
            .front()
            .is_some_and(|oldest| entry_number < number_of(oldest))
    }

    pub fn len(&self) -> usize {
        self.held().lines.len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

fn number_of(line: &Value) -> i64 {
    line.get("entry_number")
        .and_then(Value::as_i64)
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lines_are_numbered_by_the_daemon_and_kept_in_both_directions() {
        let monitor = DeviceLineMonitor::new(10);
        monitor.record("to_device", "a");
        monitor.record("from_device", "b");
        let lines = monitor.lines_since(0, 500);
        assert_eq!(lines[0]["direction"], "to_device");
        assert_eq!(lines[1]["entry_number"], 1);
        assert_eq!(lines[1]["line"], "b");
    }

    #[test]
    fn the_ring_drops_the_oldest_and_a_cursor_behind_it_is_told_so() {
        let monitor = DeviceLineMonitor::new(3);
        for index in 0..5 {
            monitor.record("from_device", &index.to_string());
        }
        assert_eq!(monitor.len(), 3);
        assert_eq!(monitor.oldest_entry_number_still_held(), 2);
        assert_eq!(monitor.newest_entry_number(), 4);
        assert!(monitor.has_fallen_out_of_the_ring(1));
        assert!(!monitor.has_fallen_out_of_the_ring(2));
    }

    #[test]
    fn an_empty_monitor_has_lost_nothing() {
        let monitor = DeviceLineMonitor::new(3);
        assert!(!monitor.has_fallen_out_of_the_ring(0));
        assert_eq!(monitor.oldest_entry_number_still_held(), 0);
        assert_eq!(monitor.newest_entry_number(), -1);
    }
}
